pyro/distributions/conjugate.py

```diff
@@ -85,3 +85,59 @@ def enumerate_support(self, expand=True):
         if expand:
             values = values.expand((-1,) + self._batch_shape)
         return values
+
+
+class GammaPoisson(TorchDistribution):
+    r"""
+    Compound distribution comprising of a gamma-poisson pair. The ``rate`` parameter
+    for the :class:`~pyro.distributions.Poisson` distribution is unknown and randomly
+    drawn from a :class:`~pyro.distributions.Gamma` distribution.
+
+    :param float or torch.Tensor concentration: shape parameter (alpha) of the Gamma
+        distribution.
+    :param float or torch.Tensor rate: rate parameter (beta) for the Gamma
+        distribution.
+    """
+
+    arg_constraints = {'concentration': constraints.positive, 'rate': constraints.positive}
+    support = Poisson.support
+
+    def __init__(self, concentration, rate, validate_args=None):
+        concentration, rate = broadcast_all(concentration, rate)
+        self._gamma = Gamma(concentration, rate)
+        super(GammaPoisson, self).__init__(self._gamma._batch_shape, validate_args=validate_args)
+
+    @property
+    def concentration(self):
+        return self._gamma.concentration
+
+    @property
+    def rate(self):
+        return self._gamma.rate
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(GammaPoisson, _instance)
+        batch_shape = torch.Size(batch_shape)
+        new._gamma = self._gamma.expand(batch_shape)
+        super(GammaPoisson, new).__init__(batch_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def sample(self, sample_shape=()):
+        rate = self._gamma.sample(sample_shape)
+        return Poisson(rate).sample()
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        return torch.lgamma(self.concentration + value) + self.concentration * self.rate.log() \
```

Comment:
Hmm I'm confused because usually for the Gamma distribution, `beta=rate` and `alpha=concentration`, whereas [this reference](http://www.math.wm.edu/~leemis/chart/UDR/PDFs/Gammapoisson.pdf) suggests we should have
```py
lgamma(rate + value) + value * concentration.log() \
- lgamma(rate) - (rate + value) * (1 + concentration).log() \
- lgamma(value + 1)
```
Are you sure you haven't swapped concentration <-> rate?

Reply:
They use some tricky notation there, which tripped me up too! 😄 

Here is their reference to the [gamma distribution](http://www.math.wm.edu/~leemis/chart/UDR/PDFs/Gamma.pdf) which should clarify the relationship with the parameters for `torch.distributions.gamma`.
 - pdf - shape parameter beta = torch.dist - concentration
 - pdf - scale parameter alpha = torch.dist - 1 / rate

Reply:
"tricky" is being generous

Reply:
Wikipedia has a good [reference](https://en.wikipedia.org/wiki/Conjugate_prior) for conjugate prior. According to it, this formula is correct. I think that we should add some reference here to avoid confusion. And I prefer to group `lgamma` operators together (called it `binomial_coef` e.g.) to make it easier to verity this formula, but this is just a nit.

Reply:
:+1: to using `_log_beta()` to group `lgamma` operations, since that's already defined in this file.

Reply:
The `lgamma` operations are grouped in `log_beta`, but note that I had to take out a `(self.concentration + value)` term in the denominator (due to `lgamma(x + 1)` in the original), so this might not be that much clearer.